### ISSUE TITLE
Uncaught TypeError: undefined is not a function

### ISSUE THREAD

**Issue by wendellmartin** [atom/markdown-preview#198]
[Enter steps to reproduce below:]

Load Atom with the following file:

---BEGIN FILE---
#facade

##Jenkins Status
BaseFacade build: [![Build Status](http://jenkins.internal.salesforce.com/buildStatus/icon?job=eapps-facade-1-BUILD-all)](http://jenkins.internal.salesforce.com/view/PDXBrewHub/job/eapps-facade-1-BUILD-all/)  
DataFacade build: [![Build Status](http://jenkins.internal.salesforce.com/buildStatus/icon?job=eapps-facade-2-BUILD-data)](http://jenkins.internal.salesforce.com/view/PDXBrewHub/job/eapps-facade-2-BUILD-data/)  
SearchFacade build: [![Build Status](https://jenkins.internal.salesforce.com/view/PDXBrewHub/job/eapps-facade-2-BUILD-search/badge/icon)](https://jenkins.internal.salesforce.com/view/PDXBrewHub/job/eapps-facade-2-BUILD-search/)  

## Architecture with links
https://docs.google.com/a/salesforce.com/file/d/0B6zmW5jk3RyJVnFiRW9DeklESVU/edit

##Project Structure

<pre>
facade
├── client
│   ├── rest
│   └── sfdc_soap_partner
├── core
│   ├── broker
│   └── model
├── deploy
│   └── data
│   └── search
├── gradle
│   └── wrapper
├── jenkins-scripts
├── jetty
├── service
│   ├── facade
│   ├── orgchart
│   ├── search
│   └── ticketing
└── util
    ├── concurrency
    ├── exception
    ├── jmx
    ├── logging
    └── query-builder

</pre>

###Gradle

  All project/directory dependencies are now properly mapped. You can clean, build, install, and eclipse from the root directory.  

  The build process and configuration has been updated so that you can "build once" and run any swimlane from a single war artifact.

To set up your environment to run the facade for a given swimlane, you will need to copy the correct jetty.sh startup script from the jetty directory.

The first time you do this, you should make a backup of the current (should be default) jetty.sh:
```
mkdir $JETTY_HOME/bin/backup
cp $JETTY_HOME/bin/jetty.sh $JETTY_HOME/backup
```
Once that is done, and whenever there is an update to the jetty.sh script (or whenever you want to switch env views). For example, to run in the 'dev' swimlane:
```
cp jetty/jetty.sh.dev $JETTY_HOME/bin/jetty.sh
```

Note that there is also a "debug" version that will pause Jetty and allow you to connect a debugger to port 9999.

Further, there is a "debug + newrelic" version that will both pause for debug and report interactions to newrelic.  For more information, see `startup.md`.

Once your env is set up, you can build and deploy from gradle as normal.  For Jenkins, these steps will not be required as they are only necessary for systems on which the jetty container will be run.

  To clean, build, and deploy the Data facade services from the root directory, use:  
```
gradle clean service:facade:build service:orgchart:build service:ticketing:build deploy:data:deploy
```

  To clean, build, and deploy the Search facade services from the root directory, use:  
```
gradle clean service:facade:build service:search:build deploy:search:deploy
```

  To start jetty, you will run the same commands used in the non-dev servers:
```
$JETTY_HOME/bin/jetty.sh start
```

The first time you do this be sure to tail the logs and ensure the env variable is set properly in the $JETTY_HOME/start.ini - if it is not set (i.e. you skipped the copy step above) you will see an exception indicating that the authentication-endpoints and other xml configuration files could not be loaded.

Note that if your $JETTY_HOME/start.ini has -Denv=<environment> (e.g. -Denv=qa) this value will override the value set in the $JETTY_HOME/bin/jetty.sh. We recommend that you remove any custom properties from the $JETTY_HOME/start.ini as we are consolidating all app-specific properties into the jetty.sh script.

###Jenkins Scripts
All scripts for gradle targets and jenkins job execution are under the `jenkins-scripts` directory.

## Installation
Steps to install facade on your MAC/Linux machine:  

1. [MAC] install Xcode and then install the command line tools  **Tip:** On Mavericks and later (10.9.x+), you can install the command line tools simply by trying to run `git` (from the command line) and OS X will download the package. (You may still need to accept user agreements and enter a password the first time you use the tools.)  
2. [ALTERNATE] you can use one of the many GIT GUI clients if you prefer. These instructions assume command line git is installed  

3. Make sure you can connect to [GIT repo][repo]:  

4. Set up GIT ssh keys:  
    Use GitHub's [guide][ssh] to setup SSH keys. Use the following command to test:
    `ssh -T git@git.soma.salesforce.com`  

5. `cd` into the directory where you want to keep the repo (e.g. `~/dev/git`)  

6. Clone the facade repo  
    `git clone git@git.soma.salesforce.com:PDXBrewHub/facade.git`  

7. check the clone  
    ```
    cd facade
    git status
    ```  
8. make sure correct version of Java is installed (version 1.6.0_65).  
    To check, run:
    `java -version`

9. Install [gradle][gradle]  
    [MAC] Recommended: copy the gradle directory to `/usr/local` (keep track of location, you'll need it for .profile)  
    [MAC] tested with gradle-1.12  

10. install [Jetty version 8][jetty]  
    **Note** that version 9 requires Java 7, which our facade does not support  
    copy jetty root directory to known location (e.g. `/usr/local`) and keep track of that location (you'll need it for the .profile)  

11. install [Apache ant][ant]  
    copy to known location (e.g. `/usr/local`) and keep track of that location (you'll need it for .profile)  

12. install [jmeter][jmeter]  
    copy to known location (e.g. `/usr/local`) and keep track of that location (you'll need it for .profile)  

13. Update system variables  
    [MAC] create or add .profile or .bash_profile - example:  
    ```Shell
    # put env vars here
    export GRADLE_HOME=/usr/local/gradle-1.12
    export JETTY_HOME=/usr/local/jetty-distribution-8.1.15.v20140411
    export JMETER_HOME=/usr/local/apache-jmeter-2.11
    export JAVA_HOME=/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home
    export JDK_HOME=/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home
    export ANT_HOME=/usr/local/apache-ant-1.9.4
    # put paths here
    export PATH=$PATH:$GRADLE_HOME/bin/
    ```  

14. Add the dev GSA cert to the Java list of trusted certificates
	No really, this is OK.  Since we are using Java 1.6, the root certificate for the GSA's cert isn't included by default in the keystore.  It's signed by VeriSign, just using a different top level cert.  For this reason you need to add the GSA cert to the keystore.  This is an example line I used to do this.  Obviously the names need to be changed to protect the innocent.  
    ```Shell
    cd /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/lib/security
    $  sudo keytool -import -alias gsadev -file /Users/wendell.martin/dev/PDXBrewHub/facade/service/facade/src/main/resources/gsa_dev.pem -keystore cacerts
    ```

	The default password for the java keystore is "changeit".  

15. Copy the correct jetty.sh script from the jetty directory. For example, to run in the 'dev' swimlane.
    ```
    mkdir $JETTY_HOME/backup
    cp %JETTY_HOME/bin/jetty.sh $JETTY_HOME/backup
    cp jetty/jetty.sh.dev $JETTY_HOME/bin/jetty.sh
    ```

### Using Homebrew

[Homebrew][brew] is a popular OS X package manager which you can use to install all the needed dependencies. While `brew` is unrelated to PDXBrewHub, the striking similarities in name choice should be reason enough to use it.

Follow steps 1 - 8 above, then continue here. Run the following command:

    brew install gradle jetty8 ant jmeter

The following environment variables should be set after installation. (You should check to make sure that the specific version numbers are up to date.)

```Shell
BREW=/usr/local/Cellar
export ANT_HOME=$BREW/ant/1.9.4
export GRADLE_HOME=$BREW/gradle/1.12
export JETTY_HOME=$BREW/jetty8/8.1.15/libexec
export JMETER_HOME=/$BEW/jmeter/2.11
export JDK_HOME=/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home
export JAVA_HOME=$JDK_HOME
 # put paths here
export PATH=$PATH:$GRADLE_HOME/bin/
```
After `brew` has finished, do the following:

Copy the correct jetty.sh script from the jetty directory. For example, to run in the 'dev' swimlane.
```
mkdir $JETTY_HOME/backup
cp %JETTY_HOME/bin/jetty.sh $JETTY_HOME/backup
cp jetty/jetty.sh.dev $JETTY_HOME/bin/jetty.sh
```

...then run:

```
gradle clean service:facade:build service:orgchart:build service:ticketing:build deploy:data:deploy
```
...to clean, build, and deploy the Data facade to your local Jetty instance.

Note if you want to change the swimlane for testing purposes, you can do so by looking for the following lines in the jetty.sh
```
#################
# FACADE-specific Jetty options (swimlane)
################
JAVA_OPTIONS+=("-Denv=dev -Xmx4g")
```
and change env=dev to the correct swimlane. For example:
```
#################
# FACADE-specific Jetty options (swimlane)
################
JAVA_OPTIONS+=("-Denv=stage -Xmx4g")
```

To ensure you're running the correct swimlane, you can grep the facade log file for the word "environment" -- when facade starts up the ApplicationListener logs an entry indicating which environment it is using.

### Overview
The GSA is our choice of search indexer.  There are three GSAs installed at the company inside the SFM network:

| Swimlane          | URL                                                | URL Type        |
|-------------------|----------------------------------------------------|-----------------|
| Dev               | https://sfm-io-gsa02.internal.salesforce.com:8443/ | admin interface |
| Prod              | https://sfm-io-gsa01.internal.salesforce.com:8443/ | admin interface |
| Disaster Recovery | https://sfm-io-gsa03.internal.salesforce.com:8443/ | admin interface |

We currently have access to the dev and disaster recovery GSA only.  We have done all our
development so far on the dev gsa.  The following folks are admins on this box:
* aswaminathan
* ksgangavi
* mcanaday
* mlynch
* ptran
* rboinepalli
* wendell.martin

If you need help connecting and configuring the GSA, but you are not one of these people,
contact them to get an account on the system.

#### Concepts
A Collection in the GSA is essentially a regex pattern match on the URL of a set of documents that segments the index
into pieces.  As an example, on the current Intranet, you can search for the entire enterprise, but you can also search
just for the employees.  This is configured with a large collection (All Enterprise), and more specific collections
(Employees).  We are using Collections to separate swimlanes.  We do this by appending a swimlane token to the end of
the URL.

By default, the swimlane token is linked to the env environment variable used by the server.  Ususally this is one of
`dev`, `qa`, `stage`, or `prod`.  This is already an established aspect of running the system.  

The swimlane token is defined by default as Answers_${env}, thus the required collections to be defined on the various
GSA servers is as follows:

| Swimlane/env | Server       | Collection Name |
|--------------|--------------|-----------------|
| dev          | sfm-io-gsa02 | Answers_dev     |
| qa           | sfm-io-gsa02 | Answers_qa      |
| stage        | sfm-io-gsa02 | Answers_stage   |
| prod         | sfm-io-gsa01 | Answers_prod    |

#### GSA Config

Since we are appending the token to the URL, we set up the Collections as follows:
0. Log into the GSA
0. Go to `Index | Collections`
0. Go to the `Create New Collection` section.
0. Enter the appropriate name for the collection (using the table above)
0. Start with an Empty `Initial Configuration`
0. Your new entry is in the list of `Current Collections`.  Click the `Edit` button beside your new Collection.
0. Enter the following into the `Include Content Matching the Following Patterns:` box:

  >regexp:Answers_{env}$

  Example: if you are setting up the collection for prod, this value should be:

  >regexp:Answers_prod$

0. Click the `Save` button at the bottom of the screen.

#### Local Configuration

If you need your own swimlane on the GSA, there are two steps.  First, pick a unique token (like your initials) to
use, for example WM.  Follow the instructions under GSA Config to create the GSA Collection (In this case, "Answers_WM").

Secondly update your Spring configuration to pass this value into your system.  
Open the file `facade/service/search/src/main/resources/app-search-config.xml`.  Locate
the `GSA` bean and look for the `collection` property.  Change this value from `Answers_${env}`
to the appropriate value that matches exactly the token you used when creating
the GSA search Collection.  In this example `Answers_WM`.

All your work now will append this token to the end of your URLs, and it will be different
from any other machine.  You now have sequestered the work you are doing into its own collection
that all other swimlanes will ignore.

#### Metadata

Since we are computing metadata, we can be relatively certain that our Answers have a specific set.  We are storing
the set of known metadata keys as an enum in `com.salesforce.webcore.eapps.facade.search.AnswerMetadataKey`.
We expect all social layer ranking-influencing metadata to be loadable from the `Answer_Metadata__c` object within
the Answers org.  This then translates directly into metadata sent to the GSA.  

Since the GSA order-by-metadata is only available for one piece of metadata (you can't by default order by two different metadata
fields or adjust comparative weights), we are using an additional `ranking` metadata which will be computed based on all
of the different available factors.  This allows us the ability to tweak the algorithm, and have the results recomputed
at next crawl.

GSA order-by-metadata is detailed in the
[GSA Documentation](http://www.google.com/support/enterprise/static/gsa/docs/admin/72/gsa_doc_set/xml_reference/request_format.html#1083224).
Our implementation will likely add the following parameter:

```
sort=meta:ranking:D:SD:::Y
```

##### Metadata Keys:
At this writing, the following are all extant and planned metadata keys:

| Name            | Values                     | Comments                          |
|-----------------|----------------------------|-----------------------------------|
| datasource      | DJC, Supportforce, ...     | The datasource type               |
| datasource_type | HowTo, Policy, ...         | the data-source specific subtype. |
| user_role       | EMPLOYEE, CONTRACTOR, ...  | The appropriate employee type.    |
| language        | en_US / jp                 | The language of the answer        |
| topic           | ESPP, Finance, ...         | The classification of the answer  |
| *like_count*    | 0-10000 (positive integer) | The number of people who have liked the Answer |
| *ranking*       | 0-? (positive integer)     | The ranking weight that the search results should be ordered by |

*italicized values are planned/proposed.*

Currently we are only proposing to use like_count to influence the ranking score.  All
other metadata pieces are used for other purposes.

## Next Steps

Continue on to the [setup doc][setup] to get façade up and running!

[ant]: http://ant.apache.org/bindownload.cgi
[brew]: http://brew.sh
[gradle]: http://www.gradle.org/installation
[jetty]: http://www.eclipse.org/jetty/
[jmeter]: http://jmeter.apache.org/download_jmeter.cgi
[repo]: https://git.soma.salesforce.com/PDXBrewHub/facade
[setup]: startup.md
[ssh]: https://help.github.com/articles/generating-ssh-keys
----END FILE---

2.  Go to Packages | Markdown Preview | Toggle Markdown
3.  Observe the error.

**Atom Version**: 0.174.0
**System**: Mac OS X 10.9.5
**Thrown From**: [markdown-preview](https://github.com/atom/markdown-preview) package, v0.120.0

### Stack Trace

Uncaught TypeError: undefined is not a function

```
At /Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:145

TypeError: undefined is not a function
  at convertCodeBlocksToAtomEditors (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:145:47)
  at /Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:41:7
  at /Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:80:16
  at module.exports (/Applications/Atom.app/Contents/Resources/app/node_modules/roaster/lib/roaster.js:66:14)
  at render (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:73:12)
  at Object.exports.toDOMFragment (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/renderer.js:30:12)
  at MarkdownPreviewView.module.exports.MarkdownPreviewView.renderMarkdownText (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/markdown-preview-view.js:246:23)
  at MarkdownPreviewView.module.exports.MarkdownPreviewView.renderMarkdown (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/markdown-preview-view.js:241:21)
  at /Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/markdown-preview-view.js:108:26
  at MarkdownPreviewView.module.exports.MarkdownPreviewView.resolveEditor (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/markdown-preview-view.js:115:16)
  at MarkdownPreviewView.module.exports.MarkdownPreviewView.attached (/Applications/Atom.app/Contents/Resources/app/node_modules/markdown-preview/lib/markdown-preview-view.js:45:21)
  at space-pen-div.View.element.attached (/Applications/Atom.app/Contents/Resources/app/node_modules/archive-view/node_modules/atom-space-pen-views/node_modules/space-pen/lib/space-pen.js:164:63)
  at space-pen-div.CustomElementPrototype.attachedCallback (/Applications/Atom.app/Contents/Resources/app/node_modules/archive-view/node_modules/atom-space-pen-views/node_modules/space-pen/lib/space-pen.js:51:55)
  at atom-pane.PaneElement.activeItemChanged (/Applications/Atom.app/Contents/Resources/app/src/pane-element.js:110:24)
  at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:82:11)
  at Pane.module.exports.Pane.setActiveItem (/Applications/Atom.app/Contents/Resources/app/src/pane.js:264:22)
  at Pane.module.exports.Pane.addItem (/Applications/Atom.app/Contents/Resources/app/src/pane.js:353:14)
  at Pane.module.exports.Pane.activateItem (/Applications/Atom.app/Contents/Resources/app/src/pane.js:327:14)
  at /Applications/Atom.app/Contents/Resources/app/src/workspace.js:518:16
  at _fulfilled (/Applications/Atom.app/Contents/Resources/app/node_modules/pathwatcher/node_modules/q/q.js:787:54)
  at self.promiseDispatch.done (/Applications/Atom.app/Contents/Resources/app/node_modules/pathwatcher/node_modules/q/q.js:816:30)
  at Promise.promise.promiseDispatch (/Applications/Atom.app/Contents/Resources/app/node_modules/pathwatcher/node_modules/q/q.js:749:13)
  at /Applications/Atom.app/Contents/Resources/app/node_modules/pathwatcher/node_modules/q/q.js:810:14
  at flush (/Applications/Atom.app/Contents/Resources/app/node_modules/pathwatcher/node_modules/q/q.js:108:17)
  at process._tickCallback (node.js:386:11)

```

### Commands

```
 22x -3:23.8 core:move-right (atom-text-editor.editor)
     -3:14.9 core:backspace (atom-text-editor.editor)
     -3:13.7 editor:newline (atom-text-editor.editor)
  3x -3:11.8 core:backspace (atom-text-editor.editor)
     -3:10.0 editor:newline (atom-text-editor.editor)
     -3:08.9 core:move-up (atom-text-editor.editor)
     -3:08.6 editor:newline (atom-text-editor.editor)
     -2:15.2 core:paste (atom-text-editor.editor)
     -2:13.9 core:move-up (atom-text-editor.editor)
  2x -1:50.6 core:save (atom-text-editor.editor)
     -0:57.3 editor:move-to-end-of-screen-line (atom-text-editor.editor)
  2x -0:56.1 editor:newline (atom-text-editor.editor)
 13x -0:44.2 core:backspace (atom-text-editor.editor)
     -0:10.6 core:move-right (atom-text-editor.editor)
     -0:07.4 core:save (atom-text-editor.editor)
  3x -0:00.0 markdown-preview:toggle (atom-text-editor.editor)
```

### Config

```json
{
  "core": {
    "themes": [
      "atom-dark-ui",
      "atom-dark-syntax"
    ]
  }
}
```

### Installed Packages

```coffee
# User
No installed packages

# Dev
No dev packages
```

/cc @atom/core

**Comment by izuzak** [comment]
Thanks for the detailed report -- I believe this is a duplicate of https://github.com/atom/markdown-preview/issues/176. It should be resolved if you upgrade to the latest version of Atom.

**Comment by wendellmartin** [comment]
Thanks for such a quick reply!  :)
